src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -2095,6 +2167,326 @@ HeapWord* ShenandoahFreeSet::allocate(ShenandoahAllocRequest& req, bool& in_new_
+  assert(req.type() == ShenandoahAllocRequest::_alloc_tlab ||
+         req.type() == ShenandoahAllocRequest::_alloc_shared ||
+         req.type() == ShenandoahAllocRequest::_alloc_cds,  "Must be");
+
+  uint start_idx = ShenandoahDirectlyAllocatableRegionAffinity::index();
+  const uint max_probes = ShenandoahDirectAllocationMaxProbes;
+  for (;;) {
+    HeapWord* obj = nullptr;
+    obj = cas_allocate_single_for_mutator<IS_TLAB>(start_idx, max_probes, req, in_new_region);
+    if (obj != nullptr) {
+      return obj;
+    }
+
+    uint next_start_index = ShenandoahDirectlyAllocatableRegionCount;
+    if (!try_allocate_directly_allocatable_regions(start_idx, false, req, obj, in_new_region, next_start_index)) {
+      if (obj != nullptr) {
+        return obj;
+      }
+
+      // No new directly allocatable region, but an existing region with sufficient memory has been found.
+      if (next_start_index != ShenandoahDirectlyAllocatableRegionCount) {
+        start_idx = next_start_index;
+      } else {
+        // try to steal from other directly allocatable regions
+        uint steal_alloc_start_idx = (start_idx + max_probes) % ShenandoahDirectlyAllocatableRegionCount;
+        uint steal_alloc_probes = ShenandoahDirectlyAllocatableRegionCount - max_probes;
+        obj = cas_allocate_single_for_mutator<IS_TLAB>(steal_alloc_start_idx,
+                                                       steal_alloc_probes,
+                                                       req,
+                                                       in_new_region);
+        if (obj != nullptr) {
+          return obj;
+        }
+
+        // No new directly allocatable region, no existing region directly allocatable region has sufficient memory.
+        return nullptr;
+      }
+    } else {
+      assert(obj != nullptr, "Must be");
+      return obj;
+    }
+  }
+}
+
+// Explicit specializations
+template HeapWord* ShenandoahFreeSet::try_allocate_single_for_mutator<true>(ShenandoahAllocRequest &req, bool &in_new_region);
+template HeapWord* ShenandoahFreeSet::try_allocate_single_for_mutator<false>(ShenandoahAllocRequest &req, bool &in_new_region);
+
+template<bool IS_TLAB>
+HeapWord* ShenandoahFreeSet::cas_allocate_in_for_mutator(ShenandoahHeapRegion* region, ShenandoahAllocRequest &req, bool &in_new_region) {
+  HeapWord* obj = nullptr;
+  size_t actual_size = req.size();
+  if (IS_TLAB) {
+    obj = region->allocate_lab_atomic(req, actual_size);
+  } else {
+    obj = region->allocate_atomic(actual_size, req);
+  }
+  if (obj != nullptr) {
+    assert(actual_size > 0, "Must be");
+    req.set_actual_size(actual_size);
+    if (pointer_delta(obj, region->bottom()) == actual_size) {
+      // Set to true if it is the first object/tlab allocated in the region.
+      in_new_region = true;
+    }
+  }
+  return obj;
+}
+
+class DirectAllocatableRegionRefillClosure final : public ShenandoahHeapRegionIterationClosure {
```

Comment:
I don't think we want to subclass ShenandoahHeapRegionIterationClosure here.  That iterates over all 2000 regions.  We only want to iterate over the 13 Directly allocatable regions.  Maybe we don't even need/want a closure iterator here.  We could just write a loop.